File: 464-Can I Win.py
class Solution(object):
    def canIWin(self, maxChoosableInteger, desiredTotal):
        """
        :type maxChoosableInteger: int
        :type desiredTotal: int
        :rtype: bool
        """
# two edge cases
        if desiredTotal <= 0:
            return True
        if maxChoosableInteger*(maxChoosableInteger+1)//2 < desiredTotal:
            return False
        
        bitmask = 0
        memo = dict()
        
        def dfs(remain, bitmask):
            if remain <= 0:
                return False
            if bitmask in memo:
                return memo[bitmask]
            for i in range(maxChoosableInteger):
                if (bitmask & (1<<i)):
                    continue
                if not dfs(remain-i-1, bitmask|(1<<i)):
                    memo[bitmask] = True
                    return True
            memo[bitmask] = False
            return False
        
        res = dfs(desiredTotal, bitmask)
        return res

File: JavaScript/464-Can I Win.js
/**
 * @param {number} maxChoosableInteger
 * @param {number} desiredTotal
 * @return {boolean}
 */
var canIWin = function(maxChoosableInteger, desiredTotal) {
    if (maxChoosableInteger*(maxChoosableInteger+1)/2 < desiredTotal){
        return false;
    }
    if (desiredTotal === 0){
        return true;
    }
    
    let picked = 0;
    memo = new Map();
    
    let helper = (remain, bitmask) =>{
        if (remain <= 0){
            return false;
        }
        
        if (memo.has(bitmask)){
            return memo.get(bitmask);
        }
        
        for (let i=1; i<=maxChoosableInteger; i++){
            if( (1<<(i-1))& bitmask ){
                continue;
            }
            if (!helper(remain-i, bitmask | (1<<(i-1)))){
                memo.set(bitmask, true);
                return true;        
            }
        }
        memo.set(bitmask, false);
        return false;
    }
    return helper(desiredTotal, picked);
};